firedrake/slate/slate.py

```diff
@@ -0,0 +1,767 @@
+algebra compiler, the SLATE Linear Algebra Compiler (SLAC),
+which interprets SLATE expressions and produces C++ kernel
+functions to be executed within the Firedrake architecture.
+
+Written by: Thomas Gibson (t.gibson15@imperial.ac.uk)
+"""
+
+import hashlib
+import operator
+
+from slate_assertions import *
+from slate_equation import SlateEquation
+
+from ufl.algorithms.domain_analysis import canonicalize_metadata
+from ufl.algorithms.map_integrands import map_integrand_dags
+from ufl.algorithms.multifunction import MultiFunction
+from ufl.algorithms.signature import compute_expression_hashdata, compute_terminal_hashdata
+from ufl.domain import join_domains, sort_domains
+from ufl.form import _sorted_integrals
+
+
+__all__ = ['CheckRestrictions', 'RemoveRestrictions',
+           'SlateIntegral', 'Tensor',
+           'Scalar', 'Vector', 'Matrix',
+           'Inverse', 'Transpose', 'UnaryOp', 'Negative', 'Positive',
+           'BinaryOp', 'TensorAdd', 'TensorSub', 'TensorMul']
+
+
+class CheckRestrictions(MultiFunction):
+    """UFL MultiFunction for enforcing cell-wise integrals to contain
+    only positive restrictions."""
+
+    expr = MultiFunction.reuse_if_untouched
+
+    def negative_restrictions(self, o):
+        raise ValueError("Cell-wise integrals must contain only positive restrictions.")
+
+
+class RemoveRestrictions(MultiFunction):
+    """UFL MultiFunction for removing any restrictions on the integrals of forms."""
+
+    expr = MultiFunction.reuse_if_untouched
+
+    def positive_restricted(self, o):
+        return self(o.ufl_operands[0])
+
+
+class SlateIntegral(object):
+    """An integral wrapping class that behaves like a UFL integral,
+    but doesn't have integrands in the same way. Its purpose is to
+    provide the appropriate integration domains and subdomain data."""
+
+    def __init__(self, ufl_integral):
+        """Constructor for the SlateIntegral class.
+
+        :arg ufl_integral: a ufl integral object."""
+
+        self._itype = ufl_integral.integral_type()
+        self._subdomain_data = ufl_integral.subdomain_data()
+        super(SlateIntegral, self).__init__()
+
+    def integral_type(self):
+        """Returns the integral type of the "integral"."""
+        return self._itype
+
+    def subdomain_data(self):
+        """Returns the subdomain data of the "integral"."""
+        return self._subdomain_data
+
+
+class Tensor(object):
+    """An abstract representation of a finite element
+    tensor in SLATE."""
+
+    # Initializing tensor id class variable for output purposes
+    tensor_id = 0
+
+    def __init__(self, arguments, coefficients, integrals):
+        """Constructor for the Tensor class.
+
+        :arg arugments: list of arguments of the associated UFL form
+        :arg coefficients: list of coefficients of the associated UFL form
+        :arg integrals: list of integrals of the associated UFL form
+        """
+
+        self.tensor_id = Tensor.tensor_id
```

Comment:
I was confused why you don't now increment `Tensor.tensor_id`, but then I see you do it individually in all the subclass constructors.  This seems wrong (?).